### PR TITLE
[warm-reboot] Add new preboot health check: verify database integrity

### PR DESCRIPTION
**Pull request by vaibhavhd**
#### What I did
Verify database integrity before proceeding with warm reboot or fast reboot.

This integrity check uses a JSON schema to validate DBs. To start with, only counters_db's table `COUNTERS_PORT_NAME_MAP` presence is verified. But, this list can advance in future.

The test logic is designed to be generic; any more databases or tables within them can be just added to schema list, and the verification logic needs no change.

#### How I did it
Added a JSON schema, and generic schema validation logic.

#### How to verify it

**Bad case:**
```
root@str-s6100-acs-2:/usr/local/bin# warm-reboot -vvv
Fri 10 Sep 2021 06:47:00 PM UTC Saving counters folder before warmboot...
Failed to validate DB's integrity. Exit code: 1. Use '-d' option to force ignore this check.
Fri 10 Sep 2021 06:47:02 PM UTC warm-reboot failure (15) cleanup ...
Fri 10 Sep 2021 06:47:03 PM UTC Cancel warm-reboot: code (0)
root@str-s6100-acs-2:/usr/local/bin#
```

SYSLOG:
```
root@str-s6100-acs-2:/usr/local/bin# tail /var/log/syslog
Sep 10 18:47:00.082206 str-s6100-acs-2 NOTICE admin: Saving counters folder before warmboot...
Sep 10 18:47:02.635871 str-s6100-acs-2 ERR /check_db_integrity.py: Database is missing tables/entries needed for reboot procedure. DB integrity check failed with:#012'COUNTERS_PORT_NAME_MAP' is a required property
Sep 10 18:47:02.646312 str-s6100-acs-2 NOTICE admin: warm-reboot failure (15) cleanup ...   <<<--------------------------------------- Check failed
Sep 10 18:47:03.685792 str-s6100-acs-2 NOTICE admin: Cancel warm-reboot: code (0)
root@str-s6100-acs-2:/usr/local/bin# 
```

**Bad case with forced ignore:**
```
root@str-s6100-acs-2:~# warm-reboot -vvvd
Fri 10 Sep 2021 08:32:04 PM UTC Saving counters folder before warmboot...
Fri 10 Sep 2021 08:32:06 PM UTC Ignoring Database integrity checks...   <<<--------------------------------------- IGNORED failure
Fri 10 Sep 2021 08:32:08 PM UTC Pausing orchagent ...
Fri 10 Sep 2021 08:32:08 PM UTC Collecting logs to check ssd health before warm-reboot...
Fri 10 Sep 2021 08:32:09 PM UTC Stopping lldp ...
Fri 10 Sep 2021 08:32:11 PM UTC Stopped lldp
Fri 10 Sep 2021 08:32:11 PM UTC Stopping nat ...
...
```

SYSLOG:
```
Sep 10 20:32:04.202216 str-s6100-acs-2 NOTICE admin: Saving counters folder before warmboot...
Sep 10 20:32:06.630111 str-s6100-acs-2 ERR /check_db_integrity.py: Database is missing tables/entries needed for reboot procedure. DB integrity check failed with:#012'COUNTERS_PORT_NAME_MAP' is a required property
Sep 10 20:32:06.638531 str-s6100-acs-2 NOTICE admin: Ignoring Database integrity checks...  <<<--------------------------------------- IGNORED failure
Sep 10 20:32:06.987446 str-s6100-acs-2 INFO admin: Checking that ASIC configuration has not changed
Sep 10 20:32:07.647537 str-s6100-acs-2 INFO admin: ASIC config unchanged, current and destination SONiC version are the same
Sep 10 20:32:08.739484 str-s6100-acs-2 NOTICE admin: Pausing orchagent ...
```

**Good case:**
```
root@str-s6100-acs-2:~# warm-reboot -vvvv
Fri 10 Sep 2021 06:56:20 PM UTC Saving counters folder before warmboot...
Fri 10 Sep 2021 06:56:29 PM UTC Pausing orchagent ...
Fri 10 Sep 2021 06:56:29 PM UTC Collecting logs to check ssd health before warm-reboot...
Fri 10 Sep 2021 06:56:29 PM UTC Stopping lldp ...
Fri 10 Sep 2021 06:56:32 PM UTC Stopped lldp
Fri 10 Sep 2021 06:56:32 PM UTC Stopping nat ...
Dumping conntrack entries failed
Warning: The unit file, source configuration file or drop-ins of nat.service changed on disk. Run 'systemctl daemon-reload' to reload units.
Fri 10 Sep 2021 06:56:33 PM UTC Stopped nat
Fri 10 Sep 2021 06:56:33 PM UTC Stopping radv ...
Fri 10 Sep 2021 06:56:33 PM UTC Stopped radv
Fri 10 Sep 2021 06:56:34 PM UTC Stopping sflow ...
Warning: The unit file, source configuration file or drop-ins of sflow.service changed on disk. Run 'systemctl daemon-reload' to reload units.
Fri 10 Sep 2021 06:56:34 PM UTC Stopped sflow
Fri 10 Sep 2021 06:56:34 PM UTC Stopping bgp ...
Fri 10 Sep 2021 06:56:38 PM UTC Stopped bgp
Fri 10 Sep 2021 06:56:38 PM UTC Stopping swss ...
Fri 10 Sep 2021 06:56:48 PM UTC Stopped swss
Fri 10 Sep 2021 06:56:48 PM UTC Initialize pre-shutdown ...
Fri 10 Sep 2021 06:56:48 PM UTC Requesting pre-shutdown ...
Fri 10 Sep 2021 06:56:49 PM UTC Waiting for pre-shutdown ...
Fri 10 Sep 2021 06:56:52 PM UTC Pre-shutdown succeeded, state: pre-shutdown-succeeded ...
Fri 10 Sep 2021 06:56:52 PM UTC Backing up database ...
Fri 10 Sep 2021 06:56:53 PM UTC Stopping teamd ...
Fri 10 Sep 2021 06:57:00 PM UTC Stopped teamd
Fri 10 Sep 2021 06:57:00 PM UTC Stopping syncd ...
Fri 10 Sep 2021 06:57:13 PM UTC Stopped syncd
Fri 10 Sep 2021 06:57:13 PM UTC Stopping all remaining containers ...
Fri 10 Sep 2021 06:57:17 PM UTC Stopped all remaining containers ...
Fri 10 Sep 2021 06:57:18 PM UTC updating ssd fw forwarm-reboot
Fri 10 Sep 2021 06:57:18 PM UTC Enabling Watchdog before warm-reboot
Watchdog armed for 180 seconds
Fri 10 Sep 2021 06:57:19 PM UTC Running x86_64-dell_s6100_c2538-r0 specific plugin...
Fri 10 Sep 2021 06:57:19 PM UTC Rebooting with /sbin/kexec -e to SONiC-OS-master.34161-dirty-20210907.143917 ...
```

Syslog:

```
Sep 10 18:56:20.820803 str-s6100-acs-2 NOTICE admin: Saving counters folder before warmboot...
Sep 10 18:56:27.269754 str-s6100-acs-2 DEBUG /check_db_integrity.py: Database integrity checks passed.    <<<--------------------------------------- Check passed
```

### CODE REVIEW

**Comment by vaibhavhd** [comment]
Tested below cases:

BRCM 202012:

Bad:
```
Aug 27 03:07:32.881889 str-s6100-acs-2 ERR /check_db_integrity.py: Database is missing tables/entries needed for reboot procedure. Config db integrity check failed with:#012'PORT' is a required property 
```
BRCM master:
Good:
```
Aug 31 18:09:27.504897 str2-dx010-acs-7 DEBUG /check_db_integrity.py: Database integrity checks passed.
```
Bad:
```
Aug 31 18:10:32.476158 str2-dx010-acs-7 ERR /check_db_integrity.py: Database is missing tables/entries needed for reboot procedure. Config db integrity check failed with:#012'ACL_TABLE' is a required property

Aug 31 18:13:37.438593 str2-dx010-acs-7 ERR /check_db_integrity.py: Database is missing tables/entries needed for reboot procedure. Config db integrity check failed with:#012'COUNTERS_PORT_NAME_MAP' is a required property
```

MLNX 202012:
Good:
```
Aug 31 19:37:46.676244 str2-msn2700-spy-1 DEBUG /check_db_integrity.py: Database integrity checks passed.
```
Bad:
```
Aug 31 19:40:34.511149 str2-msn2700-spy-1 ERR /check_db_integrity.py: Database is missing tables/entries needed for reboot procedure. Config db integrity check failed with:#012'ACL_TABLE' is a required property

Aug 31 19:43:40.777510 str2-msn2700-spy-1 ERR /check_db_integrity.py: Database is missing tables/entries needed for reboot procedure. Config db integrity check failed with:#012'COUNTERS_PORT_NAME_MAP' is a required property
```

202012 SLIM:
Good:
```
Aug 31 20:23:48.838603 str-a7050-acs-1 DEBUG /check_db_integrity.py: Database integrity checks passed.
```
Bad:
```
Aug 31 20:23:10.365894 str-a7050-acs-1 ERR /check_db_integrity.py: Database is missing tables/entries needed for reboot procedure. Config db integrity check failed with:#012'COUNTERS_PORT_NAME_MAP' is a required property
```

**Review by yxieca — APPROVED** [review]

File: scripts/fast-reboot [hunk 1]
```diff
@@ -327,6 +333,23 @@ function check_docker_exec()
     done
 }
 
+function check_db_integrity()
+{
+    if [[ "$REBOOT_TYPE" = "warm-reboot" || "$REBOOT_TYPE" = "fastfast-reboot" ]]; then
+        CHECK_DB_INTEGRITY=0
+        /usr/local/bin/check_db_integrity.py || CHECK_DB_INTEGRITY=$?
+        if [[ CHECK_DB_INTEGRITY -ne 0 ]]; then
+            if [[ x"${IGNORE_DB_CHECK}" == x"yes" ]]; then
```
Comment (neethajohn) [inline scripts/fast-reboot:342]: is there a need to even execute this script if ignore_db is set to True?

Comment (vaibhavhd) [reply scripts/fast-reboot:342]: This is in accordance with the other checks in this script (OA RESTARTCHECK test, ASIC checksum match test) which run the check, and then decide to ignore or fail.
I think this approach is still beneficial - we log the failure for future debugging and then ignore it. At any point after reboot, one can go and see what happened in the going down path.

**Review by neethajohn — APPROVED** [review]

**Comment by qiluo-msft** [comment]
This PR could not be cleanly cherry-picked to 202012. Please submit another PR.

**Pull request merged**